### PR TITLE
Add char-by-char parsing on env values

### LINKED ISSUE #31: Quotes and escape characters aren't properly handled.
Other implementations of dotenv are usually set up so that you could source the file and get the same behavior. However, this implementation does not properly handle quotes or escape characters. I would expect these to be the same.

```
DATABASE_URL=/Users/sean/Library/Application\ Support/Stuff/stuff.db
DATABASE_URL="/Users/sean/Library/Application Support/Stuff/stuff.db"
```

and I would expect this to potentially be invalid (though it wouldn't be catastrophic if it weren't):

```
DATABASE_URL=/Users/sean/Library/Application Support/Stuff/stuff.db
```

However, rust-dotenv includes the double quote characters and the backslashes. This has caused some confusion on Diesel. https://github.com/sgrif/diesel/issues/208

### PR DESCRIPTION
**Pull request by alicemaz**
this change is to allow for sh-like handling of single quote,
double quote, dollar, hash, and backslash. significantly changes
rules governing valid whitespace. adds unit tests to cover new
cases and (possibly temporarily) comments out two that should
possibly be removed.

closes #31

### CODE REVIEW

**Comment by alicemaz** [comment]
sooo I did a thing. this is a kind-of-big (and certainly breaking) change. basically after the regex captures value of a line I parse it (hopefully correctly) emulating the way sh handles quotes and escapes.

this is something we may want to discuss because I (intentionally) hard error on unquoted spaces inside values. trailing spaces/tabs should still be trimmed by the regex, and I allow for spaces/tabs preceding comments. personally I'd love to make whitespace around the equals sign an error too, but correctness on that may be too annoying to users unused to sh syntax.

I added unit tests to cover what I could think of but encourage y'all to try and break it. declined to mess wit integration tests until we figure out what to do with #25

File: src/lib.rs [hunk 1]
```diff
@@ -165,15 +258,20 @@ pub fn dotenv() -> Result<(), DotenvError> {
 
 #[test]
 fn test_parse_line_env() {
-    let input_iter = vec!["THIS_IS_KEY=hi this is value",
-                          "   many_spaces  =   wow a  maze   ",
+                          //FIXME discuss whether this behavior is desirable
+    let input_iter = vec![//"THIS_IS_KEY=hi this is value",
```
Comment (slapresta) [inline src/lib.rs]: agreed; i'm not sure why i ever thought allowing extra spaces everywhere was a feature (especially when no quoting capabilities were provided, thus making it impossible to specify keys/values with trailing spaces)

File: src/lib.rs [hunk 2]
```diff
@@ -44,6 +44,96 @@ fn named_string(captures: &Captures, name: &str) -> Option<String> {
...
+                output.push(c);
+            }
+        } else {
+            if escaped {
+                match c {
+                    '\\' => output.push(c),
+                    '\'' => output.push(c),
+                    '"' => output.push(c),
+                    '$' => output.push(c),
+                    ' ' => output.push(c),
```
Comment (slapresta) [inline src/lib.rs]: I'm pretty sure you can do something like

```
match c {
    '\\' | '\'' | '"' | '$' | ' ' => output.push(c),
    ...
}

**Comment by slapresta** [comment]
Thanks, @alicemaz!

I left a few comments. It looks good to me, nonetheless.

Here's the breaking changes this introduces (@alicemaz please correct me if I don't get it right):

* Mismatched quotes are now parsing errors (i.e. `foo'bar`, `"baz""`; use `foo\'bar`, `\"baz\"\"`, `baz\" or `"baz\""`)
* Dollar signs are now parsing errors (i.e. `$quux`; use `\$quux`)
* Spaces inside values are now parsing errors (i.e. `FOO=bar baz`; use `FOO="bar baz"`)

Personally, I find all of this reasonable. I also think the changes introduced in this pull request get us a little bit closer to having a ✨proper parser ✨ and to ✨getting rid of the scary regex thing ✨, and that's very exciting to me.

**Comment by chills42** [comment]
Sweet, those look pretty reasonable

**Comment by slapresta** [comment]
@chills42 Great. @alicemaz feel free to merge whenever you're ready!

**Comment by alicemaz** [comment]
@slapresta I'd actually defend the existence of the regex, regular grammars are safer and much easier to reason about than fiddly state-based parsers. the latter is necessary for matching delimiters but the rest it's a simple pattern to match against.

what I will do for now is change \s*=\s* to = and merge. for reference, this is the behavior I would like to conform to:

screenshot - 02132016 - 11 56 26 am

from there I'd like to move the regex out of that function so it doesn't need to be instantiated every line and so it can be tested in isolation, clean up some of the Option and Result juggling, maybe a few other things. but ideally this is all the breaking changes I plan in one go and whatever I'd do next is just implementation details.

as for what is actually breaking... there's a good rundown here http://wiki.bash-hackers.org/syntax/quoting but basically quotes are recognized by the parser and change state without being captured, and " and ' have different rules for what is allowed, eg ' are so-called strong quotes where there are _no_ escapes so '$hello' is fine but "\$hello" is necessary to get a $ literal. possibly in the future we might add variable expansion which would also conform with sh sourcing behavior but also could be substantially trickier.

(input / old output / new output)
KEY=value / value / value
KEY="value" / "value" / value
KEY=val\ ue / val\ ue / val ue
KEY=val ue / val ue / ERR

but "mismatched quotes" is better thought of as "unclosed delimiters" and the parser is a simple state machine. so '"here"' the ' are state changes and the "s are taken literally. the salient difference is "qu\"ote" is valid because the middle quote is escaped and taken literally but 'qu\'ote' is an error because strong quotes do not treat backslashes special, so the third ' is unclosed. you would want to close first like 'qu'\''ote' or just go with the simpler "qu'ote" to get the value qu'ote.

**Pull request merged**